build-tools/scripts/jenkinsfile

```diff
@@ -0,0 +1,484 @@
+                        ]],
+                        group: 'Build times', logarithmic: true, style: 'line', yaxis: 'ms'
+                    )
+                    
+                    plot(
+                        title: 'Xamarin.Forms',
+                        csvFileName: 'plot-xamarin-forms-integration-build-times.csv',
+                        csvSeries: [[
+                            displayTableFlag: true, file: 'TestResult-Timing-Xamarin.Forms-Integration.csv', inclusionFlag: 'OFF'
+                        ]],
+                        group: 'Build times', logarithmic: true, style: 'line', yaxis: 'ms'
+                    )
+                }
+
+                echo "Stage result: Plot build & test metrics: ${currentBuild.currentResult}"
+            }
+        }
+
+        stage('Upload test results') {
+            options {
+                timeout(time: 5, unit: 'MINUTES')       // Typically takes under 1 minute to upload test results
+            }
+
+            steps {
+                script {
+                    try {
+                        xunit thresholds: [
+                                failed(unstableNewThreshold: '0', unstableThreshold: '0'),
+                                skipped()                                                       // Note: Empty threshold settings per settings in the xamarin-android freestyle build are not permitted here
+                            ],
+                            tools: [
+                                NUnit2(deleteOutputFiles: true,
+                                failIfNotNew: true,
+                                pattern: 'xamarin-android/TestResult-*.xml',
+                                skipNoTestFiles: true,
+                                stopProcessingIfError: false)
+                            ]
+                    } catch (error) {
+                        echo "ERROR : Upload test results: Unexpected error: ${error}. Marking build as UNSTABLE"
+                        currentBuild.result = 'UNSTABLE'
+                    }
+                }
+
+                echo "Stage result: Upload test results: ${currentBuild.currentResult}"
+            }
+        }
+    }
+
+    post {
+        // Attempt to process & upload build results if the initial attempt to process and upload the logs failed
+        always {
+            script {
+                dir ("$XADir") {
+                    def status = 0
+
+                    if (!buildStatusUploaded) {
+                        echo "ERROR : RETRYING: Build status logs failed during 'publish packages to Azure' stage"
+
+                        if (!buildStatusProcessed) {
+                            echo "Retrying processing of build status logs"
+                            status = processBuildStatus()
+                        }
+
+                        if (status == 0) {
+                            status = publishPackages(publishBuildFilePaths)
```

Comment:
Can we have the packages published *before* building and running the unit tests?